### PR TITLE
Add a `workflow` module

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-52373]

This PR creates a new `workflow` module and moves QNode and `interfaces` into it.

File changes include:
* `pennylane/qnode.py` -> `pennylane/workflow/qnode.py`
* `pennylane/interfaces/execution.py` -> `pennylane/workflow/execution.py`
* `pennylane/interfaces/set_shots.py` -> `pennylane/workflow/set_shots.py`
* `pennylane/interfaces/jacobian_products.py` -> ` pennylane/workflow/jacobian_products.py`
* `pennylane/interfaces/autograd.py` -> `pennylane/workflow/interfaces/autograd.py`
* `pennylane/interfaces/jax.py` -> `pennylane/workflow/interfaces/jax.py`
* `pennylane/interfaces/jax_jit.py` -> `pennylane/workflow/interfaces/jax_jit.py`
* `pennylane/interfaces/tensorflow.py` -> `pennylane/workflow/interfaces/tensorflow.py`
* `pennylane/interfaces/tensorflow_autograph.py` -> `pennylane/workflow/interfaces/tensorflow_autograph.py`

A subsequent PR will flesh out the documentation changes. Only minor documentation changes are made here.

This PR also introduces deprecation warnings when anything is requested from `qml.interfaces`. It does not include warnings about things like `qml.interfaces.jax_jit.jax_jvp_execute`, but I think that is fine.

So that this PR strictly *moves* `qnode.py` and keeps its git history, we will add back in deprecation for the old qnode location in a follow up PR.

### CODE REVIEW

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -9,6 +9,8 @@ deprecations are listed below.
 Pending deprecations
 --------------------
 
+* The contents of ``qml.interfaces`` is moved inside ``qml.workflow``.
```
Comment (albi3ro) [inline doc/development/deprecations.rst:12]:
```suggestion
* The contents of ``qml.interfaces`` is moved inside ``qml.workflow``.

  - Contents moved in v0.35
  - Old import path removed in v0.36.
```

File: doc/index.rst
Comment (timmysilv) [inline doc/index.rst:1]: unrelated but are these supposed to be alphabetical? if so, we might wanna fix the `API` section (`qml_resource` appears before the Q's are done)

File: pennylane/qnn/torch.py [hunk 2]
```diff
@@ -530,7 +530,7 @@ def input_arg(self):
         return self._input_arg
 
     @staticmethod
-    def set_input_argument(input_name: Text = "inputs") -> None:
+    def set_input_argument(input_name: str = "inputs") -> None:
```
Comment (timmysilv) [inline pennylane/qnn/torch.py:533]: was something wrong with `Text`?

File: pennylane/workflow/__init__.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
```
Comment (timmysilv) [inline pennylane/workflow/__init__.py:1]:
```suggestion
# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
```

**Review by timmysilv — APPROVED** [review]
should we rename test files too? either way lgtm!

File: pennylane/qnn/torch.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/qnn/torch.py:533]: lint was complaining that `typing.Text` is deprecated.

https://docs.python.org/3/library/typing.html#typing.Text

File: doc/index.rst
Comment (albi3ro) [reply doc/index.rst:1]: and yes they are supposed to be deprecated.

**Comment by albi3ro** [comment]
> should we rename test files too? either way lgtm!

I'd like to do that in a follow up PR. I'd like to rename the "default qubit 2" in all the test names at the same time.

**Pull request merged**